### PR TITLE
feat: CrowdAction Listing & Actions

### PR DESCRIPTION
**Pull request by wizlif**

### CODE REVIEW

File: lib/application/crowdaction/spotlight/spot_light_bloc.dart [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import '../../../domain/crowdaction/crowdaction.dart';
+import '../../../domain/crowdaction/crowdaction_failures.dart';
+import '../../../domain/crowdaction/i_crowdaction_repository.dart';
+
+part 'spot_light_bloc.freezed.dart';
+part 'spot_light_event.dart';
+part 'spot_light_state.dart';
+
+@injectable
+class SpotLightBloc extends Bloc<SpotLightEvent, SpotLightState> {
```
Comment (Xazin) [inline lib/application/crowdaction/spotlight/spot_light_bloc.dart:14]: Spotlight is one word...

File: lib/application/crowdaction/subscribe/subscribe_bloc.dart [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+import '../../../domain/crowdaction/crowdaction.dart';
+import '../../../domain/crowdaction/i_crowdaction_repository.dart';
+
+part 'subscribe_bloc.freezed.dart';
+part 'subscribe_event.dart';
+part 'subscribe_state.dart';
+
+@injectable
+class SubscribeBloc extends Bloc<SubscribeEvent, SubscribeState> {
```
Comment (Xazin) [inline lib/application/crowdaction/subscribe/subscribe_bloc.dart:15]: It should be SubscriptionBloc

File: lib/application/crowdaction/subscribe/subscribe_bloc.dart [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  Stream<SubscribeState> mapEventToState(SubscribeEvent event) async* {
+    yield* event.map(
+      participate: _mapParticipateToState,
+      withDrawParticipation: _mapWithDrawParticipationToState,
+    );
+  }
+
+  Stream<SubscribeState> _mapWithDrawParticipationToState(
+      _WithDrawParticipation value) async* {
+    yield const SubscribeState.unSubscribingToCrowdAction();
```
Comment (Xazin) [inline lib/application/crowdaction/subscribe/subscribe_bloc.dart:31]: unsubscribing

File: lib/application/crowdaction/subscribe/subscribe_bloc.dart [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+    final failureOrSuccess =
+        await _crowdActionRepository.unsubscribeToCrowdAction(value.action);
```
Comment (Xazin) [inline lib/application/crowdaction/subscribe/subscribe_bloc.dart:34]: unsubscribeFromCrowdAction

File: lib/application/crowdaction/subscribe/subscribe_bloc.dart [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+    yield failureOrSuccess.fold(
+      (failure) => const SubscribeState.unSubscriptionFailed(),
```
Comment (Xazin) [inline lib/application/crowdaction/subscribe/subscribe_bloc.dart:37]: unsubscriptionFailed

File: lib/application/crowdaction/subscribe/subscribe_bloc.dart [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      (_) => const SubscribeState.unSubscriptionDone(),
```
Comment (Xazin) [inline lib/application/crowdaction/subscribe/subscribe_bloc.dart:38]: unsubscribed

File: lib/application/crowdaction/subscribe/subscribe_bloc.dart [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+  Stream<SubscribeState> _mapParticipateToState(_Participate value) async* {
+    yield const SubscribeState.subscribingToCrowdAction();
+
+    final failureOrSuccess =
+        await _crowdActionRepository.subscribeToCrowdAction(value.action);
+
+    yield failureOrSuccess.fold(
+      (failure) => const SubscribeState.subscriptionFailed(),
+      (_) => const SubscribeState.subscriptionDone(),
```
Comment (Xazin) [inline lib/application/crowdaction/subscribe/subscribe_bloc.dart:50]: subscribed

File: lib/application/crowdaction/subscribe/subscribe_event.dart [hunk 8]
```diff
@@ -0,0 +1,8 @@
+part of 'subscribe_bloc.dart';
+
+@freezed
+class SubscribeEvent with _$SubscribeEvent {
+  const factory SubscribeEvent.participate(CrowdAction action) = _Participate;
+  const factory SubscribeEvent.withDrawParticipation(CrowdAction action) =
```
Comment (Xazin) [inline lib/application/crowdaction/subscribe/subscribe_event.dart:6]: Withdraw is one word

File: lib/domain/crowdaction/i_crowdaction_repository.dart [hunk 9]
```diff
@@ -1,5 +1,16 @@
+import 'package:dartz/dartz.dart';
+
 import 'crowdaction.dart';
+import 'crowdaction_failures.dart';
 
 abstract class ICrowdActionRepository {
   Future<List<CrowdAction>> getCrowdActions({int amount = 0});
+  Future<Either<CrowdActionFailure, List<CrowdAction>>>
+      getSpotLightCrowdActions();
```
Comment (Xazin) [inline lib/domain/crowdaction/i_crowdaction_repository.dart:9]: getSpotlightCrowdActions

File: lib/domain/crowdaction/i_crowdaction_repository.dart [hunk 10]
```diff
@@ -1,5 +1,16 @@
...
+
+  Future<Either<CrowdActionFailure, Unit>> subscribeToCrowdAction(
+      CrowdAction crowdAction);
+
+  Future<Either<CrowdActionFailure, Unit>> unsubscribeToCrowdAction(
```
Comment (Xazin) [inline lib/domain/crowdaction/i_crowdaction_repository.dart:14]: unsubscribeFromCrowdAction

File: lib/presentation/shared_widgets/commitment_card.dart [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+  /// Widget for easily creating a list of CommitmentCard(s)
+  ///
+  ///[commitments] The commitments the list is made up of
+  ///
+  /// [onSelected] Callback function for when a card is selected,
+  /// returns the id of the selected commitment
+  ///
+  /// [axis] The direction of the list, defaults to [Axis.vertical]
+  const CommitmentCardList({
```
Comment (Xazin) [inline lib/presentation/shared_widgets/commitment_card.dart:19]: Can't a ListView builder replace this whole widget?

File: lib/presentation/shared_widgets/pill_button.dart [hunk 12]
```diff
@@ -41,29 +44,48 @@ class PillButton extends StatelessWidget {
...
-                  ),
-                ),
-                child: Text(text),
-              ),
+        child: _getChild(),
       ),
     );
   }
+
+  Widget _getChild() {
```
Comment (Xazin) [inline lib/presentation/shared_widgets/pill_button.dart:52]: We don't need if else statement here.

File: lib/presentation/shared_widgets/commitment_card.dart (hunk 11 above)
Comment (wizlif) [reply lib/presentation/shared_widgets/commitment_card.dart:19]: @lorenzowijtman Could you elaborate on this

**Review by Xazin — APPROVED** [review]

**Pull request merged**